subsys/bluetooth/host/iso.c

```diff
@@ -757,13 +757,15 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
 		return -ENOTCONN;
 	}
 
+	iso_conn = chan->iso;
+
 	hdr = net_buf_push(buf, sizeof(*hdr));
-	hdr->sn = sys_cpu_to_le16(sn++);
+	hdr->sn = sys_cpu_to_le16(iso_conn->iso.seq_num++);
```

Comment:
The Sequence Number cannot be simply incremented, which would have an expectation that the caller of `bt_iso_chan_send` cannot skip providing ISO data, for example, silence or media playback pause in live audio stream would mean the caller will have to call `bt_iso_chan_send` with zero length to keep up to with the sequence number.

The caller should be able to associate a data sequence w.r.t clock timeline for correct realtime reproduction at the receiver.

Reply:
> The Sequence Number cannot be simply incremented, which would have an expectation that the caller of `bt_iso_chan_send` cannot skip providing ISO data, for example, silence or media playback pause in live audio stream would mean the caller will have to call `bt_iso_chan_send` with zero length to keep up to with the sequence number.
> 
> The caller should be able to associate a data sequence w.r.t clock timeline for correct realtime reproduction at the receiver.

That is still TBD as far as I understood from our previous discussion. 

Nevertheless, this isn't fixing that part; this is just fixing the issue of a global sequence number, instead of one per channel. Once we have concluded on _how_ the number should be incremented, then we can add another PR. Consider this PR as a quick fix for a known issue.

Reply:
ok. Agreed, `hdr->sn = 0U;` would make a Controller drop all given ISO Data Packet. 

My suggestion is the caller of `bt_iso_data_send` be able to supply the Packet Sequence Number so that after a pause, new ISO Data have a Packet Sequence Number such that the Controller does not drop it (Host/Caller could use the HCI LE Read ISO TX Sync command).

Suggestions in this PR, the Packet Sequence Number is simply incremented in host alone, then after a pause/delay/drift in calling `bt_iso_data_send` all ISO Data Packet would be dropped by the Controller. This is the case in the upstream `iso_broadcast` sample, where the SDU interval is 10 ms, but `bt_iso_data_send` is called only every 1 second, hence no ISO data being transmitted on air for being with stale Packet Sequence Number.

Reply:
We definitely don't want to add the packet sequence number the API and push the responsibility to the application. This should be handled in the host (the host will have all the information to do this properly). 

To me it is still unsure if the sequence number in the host should be incremented every SDU interval, or every time a new SDU is sent to the controller, or both.

Reply:
The spec (Version 5.3, Vol 6, Part G, 2 ISOAL FEATURES) is quite clear that the sequence number shall be incremented every SDU interval.

Reply:
> The spec (Version 5.3, Vol 6, Part G, 2 ISOAL FEATURES) is quite clear that the sequence number shall be incremented every SDU interval.

Yeah, we noticed that too, but what happens if the host attempts to send 2 or more SDUs in the same interval?

Reply:
I thought the controller would generate the empty SDUs, thought in order for this to work the host would have to increment the sn based on its timing to account for those empty SDUs, I do wonder why the sn end up as part of the host though since ISOAL is part of the controller it seem to be a waste of resources to have to flow control the packets in the host.

Reply:
Controller will generate empty PDUs at real time for non-availability of SDU per SDU interval. This is determined by the use of Packet Sequence Number and optionally the timestamp supplied by the Host that shall match the SDU interval that is intended to be transmitted out. In return back to Host, Controller can be asked to give back the latest transmitted Packet Sequence Number and timestamp it maintains internally so that Host can synchronize with the Controller clock. ISOAL needs the Packet Sequence Number from Host to fragment (as necessary) the SDU into PDU and assign these PDUs with progressive Payload Count that fall into their respective ISO Intervals. 

Without the Packet Sequence Number the Isochronous nature (events occurring regularly, at equal intervals) to transmit SDUs cannot be achieved between Host and Controller, the SDUs supplied vs consumed will drift in time (sampling clock in HOst vs ISO anchor points by Controller clock)

Reply:
> I thought the controller would generate the empty SDUs, thought in order for this to work the host would have to increment the sn based on its timing to account for those empty SDUs, I do wonder why the sn end up as part of the host though since ISOAL is part of the controller it seem to be a waste of resources to have to flow control the packets in the host.

I agree. The spec is written in a way such that it sounds like the controller has it owns sequence number that it increments per SDU interval. This is based on the fact that the host can request the last sequence number sent by the controller (using the ISO TX Sync command), to sync the host sequence number with the controller sequence number, and also this paragraph:

> The sequence number shall be incremented for each SDU_Interval, whether or not an SDU was received from or sent to the upper layer.

The only reason I can find behind it, is if there's some sort of future scheduling support, i.e. if the current seq_num is 10, and the host sends a packet with seq_num 20, then the SDU would be sent in 10 SDU intervals, but such a feature is not described in any way in the core spec as far as I know.

Reply:
If the host needs to increment the sequence number for each SDU interval, it needs some sort of time. The question is then: when should the timer be started? The obvious answer to that is when it receives the e.g. CIS established event, but there's no requirements nor guarantees that this event arrives at, or is processed by, the host at the same time the controller starts it's timer for the SDU interval. If the host receives and/or processes this event e.g. 2ms after the controller, then the two respective timers would be out of sync by 2ms. 

The ISO TX Sync _may_ help here, but in case the SDU interval is 10ms, then there's still 8ms where they are synced, so more often than not the ISO TX Sync would not really sync the timers as the sequence number will be the same. 

There is the possibility of syncing based on the time_stamp instead, but that requires that the host provides the timestamp (as otherwise it will be a controller defined timestamp), or that there's some other way of synchronizing the timestamps, which also doesn't seem to be covered by the core spec. 

One alternative to keeping and incrementing a sequence number in the host, is to simple call the `HCI_LE_Read_ISO_TX_Sync` before each TX, and send the sequence number reported by that + 1, which should work in most cases, but still of course doesn't allow the host to enqueue multiple packets in a single SDU interval.